### ISSUE TITLE
false positive E0004 "match must be exhaustive" on a match of all possible u8 values

### ISSUE THREAD

**Issue by lovasoa** [intellij-rust/intellij-rust#4311]
## Environment

* **IntelliJ Rust plugin version:** 0.2.104.2132-192
* **Rust toolchain version:** 1.37.0 (eae3437df 2019-08-13) x86_64-apple-darwin
* **IDE name and version:** IC-192.6262.58
* **Operating system:** macOS 10.14.6

## Problem description
false positive E0004 "match must be exhaustive" on a match of all possible u8 values.
Found in https://github.com/lovasoa/jpeg-decoder/blob/clippy-fixes/src/marker.rs#L65-L135

## Steps to reproduce
```rust
        match n {
            0x00 => None, // Byte stuffing
            0x01 => Some(TEM),
            0x02 ..= 0xBF => Some(RES),
            0xC0 => Some(SOF(0)),
            0xC1 => Some(SOF(1)),
            0xC2 => Some(SOF(2)),
            0xC3 => Some(SOF(3)),
            0xC4 => Some(DHT),
            0xC5 => Some(SOF(5)),
            0xC6 => Some(SOF(6)),
            0xC7 => Some(SOF(7)),
            0xC8 => Some(JPG),
            0xC9 => Some(SOF(9)),
            0xCA => Some(SOF(10)),
            0xCB => Some(SOF(11)),
            0xCC => Some(DAC),
            0xCD => Some(SOF(13)),
            0xCE => Some(SOF(14)),
            0xCF => Some(SOF(15)),
            0xD0 => Some(RST(0)),
            0xD1 => Some(RST(1)),
            0xD2 => Some(RST(2)),
            0xD3 => Some(RST(3)),
            0xD4 => Some(RST(4)),
            0xD5 => Some(RST(5)),
            0xD6 => Some(RST(6)),
            0xD7 => Some(RST(7)),
            0xD8 => Some(SOI),
            0xD9 => Some(EOI),
            0xDA => Some(SOS),
            0xDB => Some(DQT),
            0xDC => Some(DNL),
            0xDD => Some(DRI),
            0xDE => Some(DHP),
            0xDF => Some(EXP),
            0xE0 => Some(APP(0)),
            0xE1 => Some(APP(1)),
            0xE2 => Some(APP(2)),
            0xE3 => Some(APP(3)),
            0xE4 => Some(APP(4)),
            0xE5 => Some(APP(5)),
            0xE6 => Some(APP(6)),
            0xE7 => Some(APP(7)),
            0xE8 => Some(APP(8)),
            0xE9 => Some(APP(9)),
            0xEA => Some(APP(10)),
            0xEB => Some(APP(11)),
            0xEC => Some(APP(12)),
            0xED => Some(APP(13)),
            0xEE => Some(APP(14)),
            0xEF => Some(APP(15)),
            0xF0 => Some(JPGn(0)),
            0xF1 => Some(JPGn(1)),
            0xF2 => Some(JPGn(2)),
            0xF3 => Some(JPGn(3)),
            0xF4 => Some(JPGn(4)),
            0xF5 => Some(JPGn(5)),
            0xF6 => Some(JPGn(6)),
            0xF7 => Some(JPGn(7)),
            0xF8 => Some(JPGn(8)),
            0xF9 => Some(JPGn(9)),
            0xFA => Some(JPGn(10)),
            0xFB => Some(JPGn(11)),
            0xFC => Some(JPGn(12)),
            0xFD => Some(JPGn(13)),
            0xFE => Some(COM),
            0xFF => None, // Fill byte
        }
```

**Comment by lovasoa** [comment]
Exhaustive integer patterns were added in rust 1.33: https://github.com/rust-lang/rust/blob/master/RELEASES.md#language-4

**Comment by TobiasNienhaus** [comment]
I've run into a similar problem and tested a bit and it seems that it only happens with u8, u16 and u32, so not with u64 and u128. I tested it using one single inclusive range spanning the whole range of the type and apparently it is also only reporting when using the literal for the max value, so using 0xFF would be reported but u8::MAX would not be. I will attach the code that I wrote to test this.

Because it might be helpful, here is also information about my IDE, etc.
IDE: CLion 2021.1.1 Build #CL-211.7142.21
Rust plugin version: 0.4.147.3871-211
OS: Windows 10 10.0
Rust version: 1.52.1 stable-x86_64-pc-windows-msvc

Code:
I added the assertions, just to be sure I didn't make any dumb mistakes.
In this code I am using hexadecimal literals, but at least 255u8 caused the same issue (I didn't test for u16, etc.) and I would imagine binary literals would cause the same.
```rust
fn main() {
    // No error reported
    match 0u8 {
        0..=u8::MAX => { }
    }
    // E0004
    match 0u8 {
        0..=0xFFu8 => { }
    }
    assert_eq!(0xFFu8, u8::MAX);

    // No error reported
    match 0u16 {
        0x0..=u16::MAX => {}
    }
    // E0004
    match 0u16 {
        0x0..=0xFFFFu16 => {}
    }
    assert_eq!(0xFFFFu16, u16::MAX);

    // No error reported
    match 0u32 {
        0x0..=u32::MAX => {}
    }
    // E0004
    match 0u32 {
        0x0..=0xFFFFFFFFu32 => {}
    }
    assert_eq!(0xFFFFFFFFu32, u32::MAX);

    // No error reported
    match 0u64 {
        0x0..=u64::MAX => {}
    }
    // No error reported
    match 0u64 {
        0x0..=0xFFFFFFFFFFFFFFFFu64 => {}
    }
    assert_eq!(0xFFFFFFFFFFFFFFFFu64, u64::MAX);

    // No error reported
    match 0u128 {
        0x0..=u128::MAX => {}
    }
    // No error reported
    match 0u128 {
        0x0..=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFu128 => {}
    }
    assert_eq!(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFu128, u128::MAX);
}
```

**Comment by ChristianLowe** [comment]
+1. In my case, I have a match statement with all 256 u8 values accounted for (all values 1:1 matched, with no range matching) with this error appearing.
